File: version2/GameServer/net_test_1/Connection.cpp
#include "Socket.h"
#include "Connection.h"

Connection::Connection( void ): m_sendingPacks(4), m_receivedPacks(8), m_delayTime(0), m_delayTimeMax(0.1f), m_sendingId(1), m_receivedId(0), m_confirmId(0) {}
Connection::~Connection( void )
{
	Close();	//	just for sure
}

void Connection::Open( const NetAddress& destination )
{
	m_destination = destination;
}

void Connection::Close( void )
{
	sx_callstack();

	//	clear sending list
	for ( int i = 0; i < m_sendingPacks.m_count; ++i )
		sx_mem_free( m_sendingPacks[i] );

	sx_callstack();

	//	clear received list
	for ( int i = 0; i < m_receivedPacks.m_count; ++i )
		sx_mem_free( m_receivedPacks[i] );

	//	clear address
	m_destination.ip = 0;
	m_destination.port = 0;
}

void Connection::SetSpeed( const int packPerSecond )
{
	m_delayTimeMax = 1.0f / packPerSecond;
}

bool Connection::Send( const void* buffer, const int size, const byte option )
{
	sx_callstack();

	if ( (size + sizeof(ConHead)) > SX_NET_CON_BUFF_SIZE )
	{
		sx_print( L"Error: Buffer size for sending with connection is out of range!\n" );
		return false;
	}

	ConPack* cp = (ConPack*)sx_mem_alloc( sizeof(ConPack) );
	cp->size = size + sizeof(ConHead);
	cp->header = (ConHead*)cp->data;
	sx_mem_copy(cp->data + sizeof(ConHead), buffer, size);

	cp->header->netId = SX_NET_ID;
	cp->header->option = option;
	cp->header->number = m_sendingId > 0 ? m_sendingId++ : (++m_sendingId)++;
	cp->header->checksum = sx_net_compute_checksum(buffer, size);

	m_sendingPacks.push_back( cp );

	return true;
}

uint Connection::Received( void* destBuffer, const int destSize )
{
	return 0;
}

void Connection::Update( class Socket* socket, const float elpsTime )
{
	m_delayTime += elpsTime;
	if ( m_delayTime < m_delayTimeMax ) return;
	m_delayTime = 0;

	sx_callstack();

	if ( m_confirmId )	//	we are waiting for received confirmation from the other side
	{
		//	search throw received list to find confirmation message
		for ( int i = m_receivedPacks.m_count - 1; i >= 0; --i )
		{
			ConPack* cp = m_receivedPacks[i];
			if ( cp->header->option & SX_NET_OPTN_CONFIRMED && cp->header->number == m_confirmId && cp->header->checksum == m_confirmChecksum)
			{
				sx_print( L"Info: Message number %d has been confirmed from other side!\n", m_confirmId );

				m_confirmId = 0;  // confirmed

				// remove the message from the list
				m_receivedPacks.remove_index(i);

				break; // we are done. so just exit the loop
			}
		}

		//	verify if the confirmation is still exist
		if ( m_confirmId )
		{
			//	check sending time out and resend the message

			//	check connection time out to close the connection
		}
	}
	else // just continue sending data
	{
		if ( m_sendingPacks.m_count )	// is there any message to send ?
		{
			ConPack* cp = m_sendingPacks[0];

			//	send message throw socket
			socket->Send(m_destination, cp->data, cp->size);

			//	verify id the message has safe send option
			if (cp->header->option & SX_NET_OPTN_SAFESEND)
			{
				m_confirmId = cp->header->number;
				m_confirmChecksum = cp->header->checksum;
			}
			else
				m_sendingPacks.remove_index(0);
		}
	}
}

void Connection::AppendReceivedMessage( class Socket* socket, const void* buffer, const int size )
{
	sx_callstack();

	//	validate message size
	if ( sx_between_i( size, sizeof(ConHead), SX_NET_CON_BUFF_SIZE ) == false )
		return;

	//	verify the package is a valid
	ConHead* ch = (ConHead*)buffer;
	{
		//	validate net id
		if ( ch->netId != SX_NET_ID )
			return;

		//	validate if message is duplicated
		if ( ch->number == m_receivedId && sx_set_hasnt( ch->option, SX_NET_OPTN_SAFESEND ) )
			return;

		//	validate data checksum
		if ( size > sizeof(ConHead) )
		{
			const byte* buf = (const byte*)buffer + sizeof(ConHead);
			if (ch->checksum != sx_net_compute_checksum(buf, size - sizeof(ConHead)))
				return;
		}
	}

	//	handle safe messages
	if ( ch->option & SX_NET_OPTN_SAFESEND )
	{
		ConHead tmp = *ch;
		tmp.option = SX_NET_OPTN_CONFIRMED;
		socket->Send( m_destination, &tmp, sizeof(ConHead) );
	}

	//	create message container and append it to the list
	if ( ch->number != m_receivedId )
	{
		ConPack* cp = (ConPack*)sx_mem_alloc( sizeof(ConPack) );
		cp->size = size;
		cp->header = (ConHead*)cp->data;
		sx_mem_copy( cp->data, buffer, size );

		//	add it to the received list
		m_receivedPacks.push_back(cp);
	}
}

File: version2/GameServer/UDP/test_unity/UdpTest/Assets/Scripts/Connection.cs
using UnityEngine;
using UnityEngine.UI;
using System.Collections;
using System.Collections.Generic;
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;

public class Connection : MonoBehaviour
{
    public Connection()
    {
        isConnected = true;
        sentNumber = 1;
        recvNumber = 0;
        confirmNumber = 0;
        confirmChecksum = 0;
        delayTime = 0;
        delayTimeMax = 0.1f;
        retryTime = 0;
        retryTimeMax = 1;
        outTime = 0;
        outTimeMax = 60;
        //net = currentNetwork;
        //m_sendingPacks = new List<NetPackage>();		//!	list of packages for send
        //m_receivedPacks = new List<NetPackage>();	//!	list of received packages
    }

    public NetBase net;

    bool isConnected;		//! specified that the connection is established
    public IPEndPoint dest;		//! connection address
    public List<NetPackage> m_sendingPacks;// = new List<NetPackage>();		//!	list of packages for send
    public List<NetPackage> m_receivedPacks;// = new List<NetPackage>();	//!	list of received packages

    ushort sentNumber;		//! used in reliability system. hold the number of message sent
    public ushort recvNumber;		//! used in reliability system. hold the number of latest message has been received

    ushort confirmNumber;	//! indicates that connection is waiting for received confirmation
    ushort confirmChecksum;	//! indicates that connection is waiting for received confirmation

    double delayTime;		//!	delay time counter
    double delayTimeMax;		//! maximum delay time

    double retryTime;		//!	retry time counter
    double retryTimeMax;		//! maximum retry time

    double outTime;			//!	time out counter
    double outTimeMax;		//! maximum time out


    void Awake()
    {
        m_sendingPacks = new List<NetPackage>();		//!	list of packages for send
        m_receivedPacks = new List<NetPackage>();	//!	list of received packages
    }

    public void Open(NetBase network, IPEndPoint destination)
    {
        dest = destination;
        net = network;
    }

    void Close()
    {
        //	clear sending list
        m_sendingPacks.RemoveAll(x => true);

        //	clear received list
        m_receivedPacks.RemoveAll(x => true);

        //	clear address
        dest = null;
    }

    public void SetSpeed(int packPerSecond)
    {
        if (packPerSecond > 0)
            delayTimeMax = 1.0f / packPerSecond;
        else
            delayTimeMax = -1;
    }

    public void SetTimeOut(double timeOut)
    {
        outTimeMax = timeOut;
    }

    public bool Send(byte[] buffer, ushort option)
    {
        NetPackage np = new NetPackage(net.NET_HEADER_SIZE, net.NET_BUFF_SIZE);
        if ((buffer.Length + net.NET_HEADER_SIZE) > net.NET_BUFF_SIZE)
        {
            return false;
        }

        np.size = buffer.Length + net.NET_HEADER_SIZE;

        Buffer.BlockCopy(buffer, 0, np.data, net.NET_HEADER_SIZE, buffer.Length);

        np.header.netId = net.NET_ID;
        np.header.option = option;
        if (sentNumber < 1) sentNumber++;
        np.header.number = sentNumber++;
        byte[] tmpBuff = net.NetHeaderToByteArray(np.header);
        np.header.checksum = (ushort)net.ComputeChecksum(buffer);
        Buffer.BlockCopy(net.NetHeaderToByteArray(np.header), 0, np.data, 0, net.NET_HEADER_SIZE);

        m_sendingPacks.Add(np);

        return true;
    }

    public int Received(out byte[] destBuffer)
    {
        if (m_receivedPacks.Count > 0)
        {
            NetPackage np = m_receivedPacks[0];

            //	remove from the list
            m_receivedPacks.Remove(np);

            //	copy package data to the destination buffer
            int size = (int)np.size - net.NET_HEADER_SIZE;

            destBuffer = new byte[size];

            Buffer.BlockCopy(np.data, net.NET_HEADER_SIZE, destBuffer, 0, size);

            return size;
        }
        else
        {
            destBuffer = null;
            return 0;
        }
    }

    public void Update()
    {
        //	check connection time out to close the connection
        outTime += Time.deltaTime;
        //if (m_outTime > m_outTimeMax)
        //{
        //    // the connection is dropped
        //    m_outTime = 0;
        //    m_connected = false;
        //    //sx_print(L"Warning: Connection time out !\n");
        //}

        if (confirmNumber > 0)	//	we are waiting for received confirmation from the other side
        {
            //	check retry time out and resend the message
            retryTime += Time.deltaTime;
            if (retryTime > retryTimeMax)
            {
                retryTime = 0;

                if (m_sendingPacks.Count > 0)
                {
                    net.socket.Send(dest, m_sendingPacks[0].data, m_sendingPacks[0].size);
                }
                else
                {
                    confirmNumber = 0;  // confirmed
                    retryTime = 0;
                    outTime = 0;
                }
            }
        }
        else // just continue sending data
        {
            if (delayTimeMax > 0)
            {
                delayTime += Time.deltaTime;
                if (delayTime < delayTimeMax) return;
                delayTime = 0;
            }

            while (m_sendingPacks.Count > 0 && confirmNumber == 0)	// is there any message to send ?
            {
                NetPackage np = m_sendingPacks[0];
                net.socket.Send(dest, np.data, np.size);

                //	verify id the message has safe send option
                if ((np.header.option & net.NET_OPTN_SAFESEND) > 0)
                {
                    confirmNumber = np.header.number;
                    confirmChecksum = np.header.checksum;
                }
                else
                {
                    m_sendingPacks.RemoveAt(0);
                }
            }
        }
    }

    public void AppendReceivedMessage(byte[] buffer, int size)
    {
        outTime = 0;
        isConnected = true;

        byte[] headerBuffer = new byte[net.NET_HEADER_SIZE];

        Buffer.BlockCopy(buffer, 0, headerBuffer, 0, net.NET_HEADER_SIZE);

        NetHeader nh = net.ByteArrayToNetHeader(buffer);

        //	we are waiting for received confirmation from the other side
        if ((nh.option & net.NET_OPTN_CONFIRMED) > 0)
        {
            if (nh.number == confirmNumber && nh.checksum == confirmChecksum)
            {
                //sx_print(L"Info: Message number %d has been confirmed from other side!\n", confirmNumber);

                confirmNumber = 0;  // confirmed
                retryTime = 0;

                // remove the message from the sending list
                m_sendingPacks.RemoveAt(0);
            }
            else
            {
                // do nothing !!
            }
        }
        else
        {
            //	replay to sender that safe message has been received
            if ((nh.option & net.NET_OPTN_SAFESEND) > 0)
            {
                NetPackage tmp = new NetPackage(net.NET_HEADER_SIZE, net.NET_BUFF_SIZE);
                tmp.header = nh;
                tmp.header.option = net.NET_OPTN_CONFIRMED;
                byte[] tmpBuf = net.NetHeaderToByteArray(tmp.header);
                Buffer.BlockCopy(tmpBuf, 0, tmp.data, 0, net.NET_HEADER_SIZE);
                net.socket.Send(dest, tmp.data, tmp.size);
            }

            //	create message container and append it to the list
            if (nh.number != recvNumber)
            {
                recvNumber = nh.number;

                NetPackage np = new NetPackage(net.NET_HEADER_SIZE, net.NET_BUFF_SIZE);
                np.size = size;

                np.header = nh;

                Buffer.BlockCopy(buffer, 0, np.data, 0, size);

                //	add it to the received list
                m_receivedPacks.Add(np);

                //sx_print( L"message received header [NO: %d OP: %d CH: %d]\n", np.header.number, np.header.option, np.header.checksum );
            }
        }
    }
}
